sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala

```diff
@@ -132,6 +135,61 @@ private[sql] object ArrowConverters {
     }
   }
 
+  private[sql] def fromPayloadIterator(iter: Iterator[ArrowPayload]): Iterator[InternalRow] = {
+    new Iterator[InternalRow] {
+      private val _allocator = new RootAllocator(Long.MaxValue)
+      private var _reader: ArrowFileReader = _
+      private var _root: VectorSchemaRoot = _
+      private var _index = 0
+
+      loadNextBatch()
+
+      override def hasNext: Boolean = _root != null && _index < _root.getRowCount
+
+      override def next(): InternalRow = {
+        val fields = _root.getFieldVectors.asScala
+
+        val genericRowData = fields.map { field =>
+          field.getAccessor.getObject(_index)
+        }.toArray[Any]
```

Comment:
@BryanCutler,

I have implemented arrow -> unsafe row conversions in:

https://github.com/icexelloss/spark/commit/8f38c15ea372c1ec9c2fefc36c0bfc3a22c3be14#diff-52cca47e7a940849b28d476ddf99d65eR575

This reuses the row object and doesn't do boxing. Hopefully it's useful to you as well?